### PR TITLE
To collection datacube

### PR DESCRIPTION
**Pull request by patrikmiksovsky95**

### CODE REVIEW

File: src/FuncSharp.Tests/Extensions/IEnumerableExtensionsTests.cs [hunk 1]
```diff
@@ -16,6 +16,19 @@ public void FirstOption()
             Assert.Equal(3.ToOption(), new List<int> { 1, 2, 3 }.LastOption());
         }
 
+        [Fact]
+        public void ToDataCollectionDataCube()
```
Comment (marektresnak) [inline src/FuncSharp.Tests/Extensions/IEnumerableExtensionsTests.cs]:
```suggestion
        public void ToCollectionDataCube()
```

File: src/FuncSharp/Extensions/IEnumerableExtensions.cs [hunk 2]
```diff
@@ -163,6 +163,123 @@ public static IOption<Exception> Aggregate(this IEnumerable<Exception> source)
...
         }
 
+        /// <summary>
+        /// Coverts the source to a new 1-dimensional data cube and aggregate values.
+        /// </summary>
+        public static DataCube1<P1, IEnumerable<TValue>> ToCollectionDataCube<T, P1, TValue>(
+            this IEnumerable<T> source,
+            Func<T, P1> p1,
+            Func<T, TValue> value,
+            Func<IEnumerable<TValue>, IEnumerable<TValue>, IEnumerable<TValue>> aggregation)
```
Comment (marektresnak) [inline src/FuncSharp/Extensions/IEnumerableExtensions.cs]: Do we need this parameter? Based on the name of the function, I would always expect it to aggregate the data. It would also make it easier to use.

File: src/FuncSharp/Extensions/IEnumerableExtensions.cs [hunk 3]
```diff
@@ -163,6 +163,123 @@ public static IOption<Exception> Aggregate(this IEnumerable<Exception> source)
             return DataCube.Create(source, p1, p2, p3, p4, p5, p6, value);
         }
 
+        /// <summary>
+        /// Coverts the source to a new 1-dimensional data cube and aggregate values.
```
Comment (marektresnak) [inline src/FuncSharp/Extensions/IEnumerableExtensions.cs]:
```suggestion
        /// Converts the source to a new 1-dimensional data cube and aggregates values.
```

File: src/FuncSharp/Extensions/IEnumerableExtensions.cs [hunk 4]
```diff
@@ -163,6 +163,123 @@ public static IOption<Exception> Aggregate(this IEnumerable<Exception> source)
...
+            Func<T, TValue> value,
+            Func<IEnumerable<TValue>, IEnumerable<TValue>, IEnumerable<TValue>> aggregation)
+        {
+            var dataCube = new DataCube6<P1, P2, P3, P4, P5, P6, IEnumerable<TValue>>();
+            foreach (var v in source)
+            {
+                dataCube.SetOrElseUpdate(p1(v), p2(v), p3(v), p4(v), p5(v), p6(v), Enumerable.Repeat(value(v), 1), aggregation);
+            }
+            return dataCube;
+        }
```
Comment (marektresnak) [inline src/FuncSharp/Extensions/IEnumerableExtensions.cs:368]: There's no changes in the TT file that generates IEnumerableExtensions.cs. If these changes were done manually, they should be done in the TT so the next time we generate the file based on the template, the changes don't get lost.

**Review by marektresnak — COMMENTED** [review]
It took me a while to understand what this does, but I can imagine the use-cases of it. I can't think of names that would describe it better though.

**Comment by siroky** [comment]
Actually, I think the best approach would be to have the methods as follows:

- ToDataCube with optional aggregation parameter. If not specified the following function should be used: (a, b) => b That is the current behavior of ToDataCube.
- ToCollectionDataCube which will concat it to collection using the ToDataCube method above.

File: src/FuncSharp/Extensions/IEnumerableExtensions.cs (hunk 4 above)
Comment (patrikmiksovsky95) [reply src/FuncSharp/Extensions/IEnumerableExtensions.cs:368]: even the original ToDataCube is not there 🤷

**Comment by patrikmiksovsky95** [comment]
We cannot have it as optional parameter as the return value will be with IEnumerable<TValue> which is changing structure and it would be breaking change + compilator won't know which implementation to use as number of parameters will be same as optional wont be counted and return type will be different, so I will do it as ToDataCube with required parameter so caller side will by themself call one or second implementation and ToCollectionDataCube will do concating with new method

**Comment by siroky** [comment]
@patrikmiksovsky95 you're right, in that case, it can be:

- A new method with mandatory aggregation function, `TResult` as another parameter and with two functions:
  - `Tvalue => TResult` so that you can initialize the result based on single element 
  - Aggregation function `(TResult, TValue) => TResult` to aggregate another value into the result.
- The current ToDataCube can utilize the method above by specifying `TResult` same as `TValue` and using:
  - `a => a` as the initialization functoin
  - `(a, b) => b` as the aggregation function.
- The new method ToCollectionDataCube can utilize the first method by specifying `TResult` as collection of `TValue` and:
  - `a => a.ToEnumerable()` as the initialization functoin
  - `(a, b) => a.Concat(b)` as the aggregation function.
  
  I think this would nicely abstract it out. Also make sure you update the TT files, do not edit the C# sources, those should get generated.

File: src/FuncSharp/Extensions/IEnumerableExtensions.tt [hunk 5]
```diff
@@ -87,19 +87,55 @@ namespace FuncSharp
         {
             return source.SingleOption().OrElse(_ => source.FirstOption().Map<Exception>(e => new AggregateException(source)));
         }
+        <#  for (var i = 1; i <= MaxCubeArity(); i++) { #>
```
Comment (patrikmiksovsky95) [inline src/FuncSharp/Extensions/IEnumerableExtensions.tt:90]: @siroky can you please have a look ?

File: src/FuncSharp/Extensions/IEnumerableExtensions.tt [hunk 6]
```diff
@@ -87,19 +87,55 @@ namespace FuncSharp
...
         {
-            return DataCube.Create(source, <#= List(i, x => "p" + x) #>, value);
+            return ToDataCube<T, <#= Types(i, name: "P") #>, TValue, TValue>(source, <#= List(i, x => "p" + x) #>, value, a => a, (a, b) => b);
         }
 <#  } #>
+
+<#  for (var i = 1; i <= MaxCubeArity(); i++) { #>
+
+        /// <summary>
+        /// Converts the source to a new <#= i #>-dimensional data cube.
```
Comment (siroky) [inline src/FuncSharp/Extensions/IEnumerableExtensions.tt]: Let's describe further that in case of collisions, it aggregates the values to a collection.

File: src/FuncSharp/Extensions/IEnumerableExtensions.tt [hunk 7]
```diff
@@ -87,19 +87,55 @@ namespace FuncSharp
         {
             return source.SingleOption().OrElse(_ => source.FirstOption().Map<Exception>(e => new AggregateException(source)));
         }
+        <#  for (var i = 1; i <= MaxCubeArity(); i++) { #>
+
+        /// <summary>
+        /// Converts the source to a new <#= i #>-dimensional data cube and aggregates values.
```
Comment (siroky) [inline src/FuncSharp/Extensions/IEnumerableExtensions.tt]: ... and aggregates the values in case of conflicting positions.

File: src/FuncSharp/Extensions/IEnumerableExtensions.tt [hunk 8]
```diff
@@ -87,19 +87,55 @@ namespace FuncSharp
...
+        /// </summary>
+        public static <#= DataCubeType(i, valueType: "TResult") #> ToDataCube<T, <#= Types(i, name: "P") #>, TValue, TResult>(
+            this IEnumerable<T> source,
+<#= Lines(i, x => Indent(12) + "Func<T, P" + x + "> p" + x, separator: ",") #>,
+            Func<T, TValue> value,
+            Func<TValue, TResult> initialization,
+            Func<TResult, TResult, TResult> aggregation)
+        {
+            var dataCube = new <#= DataCubeType(i,valueType: "TResult") #>();
```
Comment (siroky) [inline src/FuncSharp/Extensions/IEnumerableExtensions.tt]: missing space after comma `(i,valueType:`

File: src/FuncSharp/Extensions/IEnumerableExtensions.tt [hunk 9]
```diff
@@ -87,19 +87,55 @@ namespace FuncSharp
...
         /// <summary>
-        /// Coverts the source to a new <#= i #>-dimensional data cube.
+        /// Converts the source to a new <#= i #>-dimensional data cube.
         /// </summary>
         public static <#= DataCubeType(i) #> ToDataCube<T, <#= Types(i, name: "P") #>, TValue>(
             this IEnumerable<T> source,
 <#= Lines(i, x => Indent(12) + "Func<T, P" + x + "> p" + x, separator: ",") #>,
             Func<T, TValue> value)
         {
-            return DataCube.Create(source, <#= List(i, x => "p" + x) #>, value);
```
Comment (siroky) [inline src/FuncSharp/Extensions/IEnumerableExtensions.tt:100]: If DataCube.Create still exists, maybe it's more performant than this, please check that and if that's the case, I'd revert this change. We don't necessarily need to use ToDataCube here if it's not worth it.

File: src/FuncSharp/Extensions/IEnumerableExtensions.tt [hunk 10]
```diff
@@ -87,19 +87,55 @@ namespace FuncSharp
...
+
+        /// <summary>
+        /// Converts the source to a new <#= i #>-dimensional data cube and aggregates values.
+        /// </summary>
+        public static <#= DataCubeType(i, valueType: "TResult") #> ToDataCube<T, <#= Types(i, name: "P") #>, TValue, TResult>(
+            this IEnumerable<T> source,
+<#= Lines(i, x => Indent(12) + "Func<T, P" + x + "> p" + x, separator: ",") #>,
+            Func<T, TValue> value,
+            Func<TValue, TResult> initialization,
+            Func<TResult, TResult, TResult> aggregation)
```
Comment (siroky) [inline src/FuncSharp/Extensions/IEnumerableExtensions.tt]: The aggregation method should be of type `Func<TResult, TValue, TResult>` to avoid possibly useless initialization which is visible here https://github.com/siroky/FuncSharp/pull/92/files#diff-c66b6a7351d5ead614565a2b81008b44155ec1dd083fa74cef8d3e7421311495R368 with the Enumerable.Repeat. The initialization should be used only first time for the first element. And all the others should be aggregated without the need of initialization.

Maybe this will require another variant of SetOrElseUpdate, but that could be helpful as well.

File: src/FuncSharp/Extensions/IEnumerableExtensions.cs [hunk 11]
```diff
@@ -160,9 +283,92 @@ public static IOption<Exception> Aggregate(this IEnumerable<Exception> source)
             Func<T, P6> p6,
             Func<T, TValue> value)
         {
-            return DataCube.Create(source, p1, p2, p3, p4, p5, p6, value);
+            return ToDataCube<T, P1, P2, P3, P4, P5, P6, TValue, TValue>(source, p1, p2, p3, p4, p5, p6, value, a => a, (a, b) => b);
+        }
+
+
```
Comment (siroky) [inline src/FuncSharp/Extensions/IEnumerableExtensions.cs]: Multiple blank lines, needs to be fixed in the TT file

**Review by siroky — CHANGES_REQUESTED** [review]

File: src/FuncSharp/Extensions/IEnumerableExtensions.tt (hunk 9 above)
Comment (patrikmiksovsky95) [reply src/FuncSharp/Extensions/IEnumerableExtensions.tt:100]: It internally does the same thing, create, foreach, set

File: src/FuncSharp/Extensions/IEnumerableExtensions.tt (hunk 10 above)
Comment (patrikmiksovsky95) [reply src/FuncSharp/Extensions/IEnumerableExtensions.tt]: @siroky there is still some initialization as current C# in funcsharp doesnt have append yet.
https://github.com/siroky/FuncSharp/pull/92/files#diff-c66b6a7351d5ead614565a2b81008b44155ec1dd083fa74cef8d3e7421311495R297
